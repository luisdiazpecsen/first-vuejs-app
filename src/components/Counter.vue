<template>
    <h2>{{ customTitle }}</h2>
    <p>{{ counter }}<sup>2</sup> = {{ squareCounter }}</p>
    <p data-testid="counter">{{ counter }}</p>

    <div>
        <button data-testid="decrease" @click="decrease">-1</button>
        <button data-testid="increase" @click="increase">+1</button>
    </div>
</template>

<script>
export default {
    name: "Counter",
    // props: ["counter-title", "start"],
    props: {
        counterTitle: String,
        start: {
            type: Number,
            required: false,
            default: 1,
            validator: function (value) {
                return value > 0
            },
        },
    },
    data() {
        return {
            counter: this.start,
        }
    },
    methods: {
        getSquareValue() {
            return this.counter * this.counter
        },
        increase() {
            this.counter++
        },
        decrease() {
            this.counter--
        },
    },
    computed: {
        squareCounter() {
            return this.counter * this.counter
        },
        customTitle() {
            return this.counterTitle || "Counter"
        },
    },
}
</script>

<style>
button {
    background-color: #64bb87;
    border-radius: 5px;
    border: 1px solid white;
    color: white;
    cursor: pointer;
    margin: 0 5px;
    padding: 5px 15px;
    transition: 0.3s ease-in-out;
}

button:hover {
    background-color: #5aa67b;
    transition: 0.3s ease-in-out;
}
</style>