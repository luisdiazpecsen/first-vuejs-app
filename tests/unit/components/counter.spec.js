import { shallowMount } from '@vue/test-utils'
import Counter from '@/components/Counter.vue'

describe('Counter Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(Counter)
    })

    // SNAPSHOT
    test('should match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    })


    // HTML TAG
    test("should have 'Counter' as the h2's default value", () => {
        expect(wrapper.find('h2').exists()).toBeTruthy()

        const hValue = wrapper.find('h2').text()
        expect(hValue).toBe('Counter')
    })


    // FINDALL
    test("should have 1 as the p's default value", () => {
        // const  pTags = wrapper.findAll('p')
        // expect(pTags[1]).not.toBeUndefined()

        // const value = pTags[1].text
        // expect(value).toBe('1')

        const pTag = wrapper.find('[data-testid="counter"]')
        expect(pTag.exists()).toBeTruthy()

        const value = pTag.text()
        expect(value).toBe('1')
    })


    // SIMULATE EVENTS
    test("should increase and decrease the counter's value", async () => {
        // const increaseBtn = wrapper.find('[data-testid="increase"]'),
        //     decreaseBtn = wrapper.find('[data-testid="decrease"]')

        const [decreaseBtn, increaseBtn] = wrapper.findAll('button')

        expect(increaseBtn.exists()).toBeTruthy()
        expect(decreaseBtn.exists()).toBeTruthy()

        // Click the increase button
        await increaseBtn.trigger('click')
        // Click the decrease button
        await decreaseBtn.trigger('click')
        await decreaseBtn.trigger('click')

        const value = wrapper.find('[data-testid="counter"]').text()
        expect(value).toBe('0')
    })


    // PROPERTIES
    test('should set the default value', () => {
        const start = wrapper.props('start'),
            value = wrapper.find('[data-testid="counter"]').text()
        expect(value).toBe(start.toString())
        expect(Number(value)).toBe(start)
    })


    // SEND PROPERTIES
    test('should show the title property', () => {
        const title = 'Hola mundo'
        const wrapper = shallowMount(Counter, {
            props: {
                counterTitle: title
            }
        })
        expect(wrapper.find('h2').text()).toBe(title)
    })
})