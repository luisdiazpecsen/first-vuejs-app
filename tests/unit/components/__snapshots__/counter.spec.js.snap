// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component should match with snapshot 1`] = `
<h2>Counter</h2>
<p>1<sup>2</sup> = 1</p>
<p data-testid="counter">1</p>
<div><button data-testid="decrease">-1</button><button data-testid="increase">+1</button></div>
`;
